Adding Deep Immutability to 3.12

matajoh opened the pull request:
This draft PR is to give a chance to view the diff between 3.12 and just adding the deep immutability from veronapy. The key difference is the use of the lowest bit of `ob_refcnt` to indicate immutability (so as to not impact object header size).

<!-- readthedocs-preview cpython-previews start -->
----
📚 Documentation preview 📚: https://cpython-previews--51.org.readthedocs.build/

<!-- readthedocs-preview cpython-previews end -->

xFrednet commented on Include/internal/pycore_immutability.h:
```diff
@@ -0,0 +1,22 @@
+#ifndef Py_FREEZE_H
```
Should this header maybe be called `pycore_freeze.h`? Same with the .c file?

matajoh replied:
As discussed I think we'll keep it as is for now to avoid any confusion with the deep freeze functionality, but wait and see what feedback we receive.

mjp41 commented on Doc/c-api/typeobj.rst:
```diff
@@ -2591,9 +2591,9 @@ Slot Type typedefs
    The purpose of this function is to separate memory allocation from memory
    initialization.  It should return a pointer to a block of memory of adequate
    length for the instance, suitably aligned, and initialized to zeros, but with
-   :c:member:`~PyObject.ob_refcnt` set to ``1`` and :c:member:`~PyObject.ob_type` set to the type argument.  If
```
Is this out of date?

matajoh committed:
Proposed code changes for [PEP XXX](link).

Signed-off-by: Matthew A Johnson <matjoh@microsoft.com>

xFrednet commented on Include/internal/pycore_runtime_init_generated.h:
```diff
@@ -554,10 +554,12 @@ extern "C" {
     INIT_STR(open_br, "{"), \
     INIT_STR(percent, "%"), \
     INIT_STR(shim_name, "<shim>"), \
+    INIT_STR(startswith, "startswith"), \
```
Are these new interned strings needed or just nice to haves?

matajoh replied:
This basically functions like a string constant, and given that I need to use these strings every time I walk a function I want to streamline their use. They are nice to have, but definitely justified.

xFrednet commented on Include/internal/pycore_dict.h:
```diff
@@ -24,14 +24,24 @@ typedef struct {
+    PyObject *_me_value; /* This field is only meaningful for combined tables */
 } PyDictUnicodeEntry;
 
+#define _PyDictEntry_IsImmutable(entry) (((uintptr_t)((entry)->_me_value)) & 0x1)
```
NIT: Do we maybe want to name the bit flag instead of hard coding a 0x1?

matajoh replied:
Good point!

xFrednet commented on Include/internal/pycore_dict.h:
```diff
@@ -24,14 +24,24 @@ typedef struct {
+#define _PyDictEntry_SetImmutable(entry) ((entry)->_me_value = (PyObject*)((uintptr_t)(entry)->_me_value | 0x1))
+#define _PyDictEntry_Hash(entry) ((entry)->me_hash)
+#define _PyDictEntry_Key(entry) ((entry)->me_key)
+#define _PyDictEntry_Value(entry) ((PyObject*)((((uintptr_t)((entry)->_me_value)) >> 1) << 1))
```
Question: Is there a reason why this is using two bit shifts opposed to the flag?

Could it be instead something like this? `(entry)->_me_value & (~DICT_VALUES_IMMUTABILITY_FLAG)`

matajoh replied:
yeah I think that is probably fine.

xFrednet commented on Include/internal/pycore_dict.h:
```diff
@@ -24,14 +24,24 @@ typedef struct {
+#define _PyDictEntry_Key(entry) ((entry)->me_key)
+#define _PyDictEntry_Value(entry) ((PyObject*)((((uintptr_t)((entry)->_me_value)) >> 1) << 1))
+#define _PyDictEntry_SetValue(entry, value) ((entry)->_me_value = value)
+#define _PyDictEntry_IsEmpty(entry) ((entry)->_me_value == NULL)
```
Could it happen that an empty dictionary is made immutable and then would fail this check even if it's empty?

matajoh replied:
yeah this needs to involve a bit mask.

xFrednet commented on Include/object.h:
```diff
@@ -78,6 +78,25 @@ whose size is determined when the object is allocated.
+#define _Py_IMMUTABLE_FLAG 0x4000000000
+#define _Py_IMMUTABLE_SCC_FLAG 0x8000000000
+#else
+#define _Py_REFCNT_MASK 0x3FFFFFFF
+#define _Py_IMMUTABLE_MASK 0xC0000000
+#define _Py_IMMUTABLE_FLAG 0x40000000
+#define _Py_IMMUTABLE_SCC_FLAG 0x80000000
```
I always don't know with these constants if it's nicer to hardcode them or construct them in the file. Is there a preference or reason to prefer either of these options?
```
#define _Py_REFCNT_MASK 0x3FFFFFFF
#define _Py_IMMUTABLE_FLAG 0x40000000
#define _Py_IMMUTABLE_SCC_FLAG 0x80000000
#define _Py_IMMUTABLE_MASK (_Py_IMMUTABLE_FLAG | _Py_IMMUTABLE_SCC_FLAG)
```

matajoh replied:
I think your version is probably better.

xFrednet commented on Include/internal/pycore_object.h:
```diff
@@ -91,6 +92,17 @@ static inline void _Py_ClearImmortal(PyObject *op)
+}
+#define _Py_SetImmutable(op) _Py_SetImmutable(_PyObject_CAST(op))
+
+#define Py_CHECKWRITE(op) ((op) && (!_Py_IsImmutable(op) || _Py_IsFinalizing()))
```
Is the need for `_Py_IsFinalizing()` here documented somewhere? Mainly asking for the future when the reason is no longer fresh in our minds.

matajoh replied:
we should 100% document why this is there, great suggestion

xFrednet commented on Objects/abstract.c:
```diff
@@ -34,6 +34,17 @@ null_error(void)
     return NULL;
 }
 
+static PyObject *
+immutable_error(void)
+{
+    PyThreadState *tstate = _PyThreadState_GET();
+    if (!_PyErr_Occurred(tstate)) {
+        _PyErr_SetString(tstate, PyExc_NotWriteableError,
+                         "cannot modify immutable instance");
+    }
+    return NULL;
+}
+
```
Could it be helpful to attach the object that is immutable and caused the exception? I'm considering if it would make debugging easier. There is also a question if this would add more burden on maintaining and adding the immutability code.

Do you know what the norm is for other Python exceptions? Do they usually provide more data or mainly the message and stacktrace?

matajoh replied:
this is not the "real" exception, but a special one for when the normal exception process cannot be used. That said, there is an argument for finding a way to include that information if possible. Let me look at this again, this code is quite old now.

matajoh replied:
So after looking at this in more detail, I think what is here (i.e. message and stacktrace) is the standard practice. That said, let's see what the feedback is from the core devs on this one.

xFrednet commented on Tools/gdb/libpython.py:
```diff
@@ -782,7 +782,8 @@ def iteritems(self):
             ep = entries[i]
-            pyop_value = PyObjectPtr.from_pyobject_ptr(ep['me_value'])
+            value_ptr = ep['_me_value'].reinterpret_cast(gdb.lookup_type('uintptr_t'))
+            pyop_value = PyObjectPtr.from_pyobject_ptr((value_ptr >> 1) << 1)
```
```suggestion
            // Clear immutability flag
            pyop_value = PyObjectPtr.from_pyobject_ptr((value_ptr >> 1) << 1)
```

xFrednet reviewed (commented):
My comments can be quite nit-picky, so please don't hesitate to push back on them :)

I'm not done reviewing everything yet, but from what I've seen thus far: Excellent job! It's amazing that you managed to get all of this together, well done :D
This review is still incomplete.  From all I've read, excellent job!

xFrednet commented on Python/immutability.c:
```diff
@@ -0,0 +1,376 @@
+        return NULL;
+    }
+
+    Py_INCREF(item);
```
Question:
Why is the item RC increased here? Shouldn't `PyList_GetItem` already increased the RC? And if not, should this `Py_INCREF(item);` be moved before resizing the slice in line 22?

matajoh replied:
PyList_GetItem() returns a borrowed ref, so the incref is needed. It should probably be moved to before the setslice, though, that's a good point.

mjp41 replied:
If you move it before, then a failure of SetSlice will leak?

mjp41 replied:
But SetSlice can remove the RC, so I think you need to add a DecRef for the failure case and move earlier.

xFrednet commented on Python/immutability.c:
```diff
@@ -0,0 +1,376 @@
+/**
+ * Special function for walking the reachable graph of a function object.
+ *
+ * This is necessary because the function object has a pointer to the global
+ * object, and this is problematic because freezing any function will make the
+ * global object immutable, which is not always the desired behaviour.
```
Is this comment still up to date with the semantics for function freezing we settled on?

This is a small suggestion for reformulation if we keep the current comment.

```suggestion
 * This is necessary because the function object has a pointer to the global
 * object. Freezing it like any other object would also make the
 * global object immutable, which is usually the undesired behaviour.
```

matajoh replied:
yup out of date comment

xFrednet commented on Python/immutability.c:
```diff
@@ -0,0 +1,376 @@
+        _Py_DECLARE_STR(namedtuple, "namedtuple");
+        _Py_DECLARE_STR(startswith, "startswith");
+        PyObject* res = PyObject_CallMethodOneArg(f->func_module, &_Py_STR(startswith), &_Py_STR(namedtuple));
+        if(Py_IsTrue(res)){
+            // namedtuple creates a fake module, which cannot be imported
+            Py_RETURN_NONE;
+        }
```
Question:
Is it correct that this just returns and doesn't attempt to freeze the rest of the function?

I would have expected a structure that attempts to get the module and just continues if this is not possible. If this is intended, could you add an explanatory comment?

matajoh replied:
so, it only is OK in the very, very specific circumstance that the thing being frozen is a namedtuple, which is why I test for that. Otherwise this is an error state. The module should always be importable, and that not being true indicates a deeper issue.

xFrednet commented on Python/immutability.c:
```diff
@@ -0,0 +1,376 @@
+                }
+            }
+
+            if(check_globals && PyUnicode_Check(value)){
```
This `if`-block was a bit hard to understand. Is my current understanding correct:

When a function accesses `globals()` it might use a string to get a value out of global. For example, the `"hej"` in this example `globals()["hej"]`. This check freezes and transfers the value of `"hej"` into the `frozen_globals`.

If this understanding is correct, could you maybe add a small comment?

matajoh replied:
yup, that's right. I'll add a comment.

xFrednet commented on Python/immutability.c:
```diff
@@ -0,0 +1,376 @@
+
+PyObject* _Py_Freeze(PyObject* obj)
+{
+    if(_Py_IsImmutable(obj) && _Py_IsImmutable(Py_TYPE(obj))){
```
Question: 
Do you have an example how it can happen that the object is immutable, but the type is mutable?

matajoh replied:
There was an example previously, but it is defunct now. I don't think this is possible anymore. Let me think on it.

xFrednet commented on Python/immutability.c:
```diff
@@ -0,0 +1,376 @@
+        _Py_SetImmutable(item);
+
+
+        if(is_c_wrapper(item)) {
+            // C functions are not mutable, so we can skip them.
+            continue;
+        }
```
Could we mark all c-wrapper types as immutable by default? I remember that we talked about this some time ago and that this was planned for "Phase 2" before we changed the order and approach.

matajoh replied:
Yeah I should look at this again too. It may be that we can just mark them immutable in the normal way and call it a day instead of leaving them around and requiring this check.

xFrednet commented on Lib/test/test_freeze.py:
```diff
@@ -0,0 +1,473 @@
+        self.assertFalse(isimmutable(global_canary))
+        self.assertRaises(NotWriteableError, d)
+
+    def test_globals_copy(self):
+        def f():
+            global global0
+            return global0
```
The special handling for globals uses `PyDict_SetItem`, this means that multiple references to the same value will overwrite the previous value of `frozen_globals`. This should be totally fine, since the old and new value should be identical.

Could modify this test to reference `global0` multiple times, something like this:
```suggestion
    def test_globals_copy(self):
        def f():
            global global0
            ref_1 = global0
            ref_2 = global0
            return global0
```

I mainly want to ensure that the RC and so on remains correct after this :)

The same could be done for builtins, but if this works the others should as well.

xFrednet reviewed (approved):
Very nicely done! This is mainly a collection of questions, which probably come from me missing some context.

mjp41 commented on Python/immutability.c:
```diff
@@ -0,0 +1,376 @@
+                    }
+                }
+            }
+        }
```
I don't understand what this is doing?  Could you explain an example that this is catching that is a problem?

matajoh committed:
Addressing PR comments

Signed-off-by: Matthew A Johnson <matjoh@microsoft.com>

matajoh committed:
Cleaning up the immutability code

Signed-off-by: Matthew A Johnson <matjoh@microsoft.com>

matajoh committed:
deque working

Signed-off-by: Matthew A Johnson <matjoh@microsoft.com>

matajoh committed:
Adding write barriers for array

Signed-off-by: Matthew A Johnson <matjoh@microsoft.com>

mjp41 commented on Python/immutability.c:
```diff
@@ -0,0 +1,403 @@
+                    Py_DECREF(f_stack);
+                    return NULL;
+                }
+            }else if(PyDict_Contains(module_dict, name)){
+                PyObject* value = PyDict_GetItem(module_dict, name);
+
+                _PyDict_SetKeyImmutable((PyDictObject*)module_dict, name);
+
+                if(!_Py_IsImmutable(value)){
+                    if(push(frontier, value)){
+                        goto nomemory;
+                    }
+                }
+            }
```
I couldn't work out how to exercise this code.  It doesn't appear to be hit by `test_freeze`.  Should we be able to get coverage for this?  I had an experiment with the REPL and a few modules, but this code path didn't get exercised.

matajoh committed:
Cleaning up function freezing + more tests

Signed-off-by: Matthew A Johnson <matjoh@microsoft.com>

matajoh committed:
Immutability for blake2 + test_freeze module

Signed-off-by: Matthew A Johnson <matjoh@microsoft.com>

matajoh committed:
Fixing some holes in object/typeobject

Signed-off-by: Matthew A Johnson <matjoh@microsoft.com>

mjp41 committed:
Remove immutable keys from dictionary as no longer required

matajoh committed:
Dealing with buffer immutability

Signed-off-by: Matthew A Johnson <matjoh@microsoft.com>

matajoh committed:
_ctypes immutability

Signed-off-by: Matthew A Johnson <matjoh@microsoft.com>

matajoh committed:
_decimal

Signed-off-by: Matthew A Johnson <matjoh@microsoft.com>

matajoh committed:
Adding a NotFreezable type

Signed-off-by: Matthew A Johnson <matjoh@microsoft.com>

matajoh committed:
_io

Signed-off-by: Matthew A Johnson <matjoh@microsoft.com>

matajoh committed:
Fixing some broken tests

Signed-off-by: Matthew A Johnson <matjoh@microsoft.com>

matajoh committed:
Fixing the name of NotWritableError

Signed-off-by: Matthew A Johnson <matjoh@microsoft.com>

matajoh committed:
_multiprocessing _sqlite3

Signed-off-by: Matthew A Johnson <matjoh@microsoft.com>

matajoh committed:
cjkcodecs

Signed-off-by: Matthew A Johnson <matjoh@microsoft.com>

matajoh committed:
_abc

Signed-off-by: Matthew A Johnson <matjoh@microsoft.com>

matajoh committed:
_asyncio

Signed-off-by: Matthew A Johnson <matjoh@microsoft.com>

matajoh committed:
Fixing import error

Signed-off-by: Matthew A Johnson <matjoh@microsoft.com>

matajoh committed:
bz2

Signed-off-by: Matthew A Johnson <matjoh@microsoft.com>

matajoh committed:
_collections _csv

Signed-off-by: Matthew A Johnson <matjoh@microsoft.com>

matajoh committed:
Moving Py_Freeze to public API

Signed-off-by: Matthew A Johnson <matjoh@microsoft.com>

xFrednet committed:
xxlimited, xxsubtype, zlib

Signed-off-by: Matthew A Johnson <matjoh@microsoft.com>

matajoh committed:
_dbm etree

Signed-off-by: Matthew A Johnson <matjoh@microsoft.com>

matajoh committed:
immutable module

Signed-off-by: Matthew A Johnson <matjoh@microsoft.com>

matajoh committed:
All tests passing again

Signed-off-by: Matthew A Johnson <matjoh@microsoft.com>

mjp41 commented on Python/immutability.c:
```diff
@@ -393,66 +368,89 @@ PyObject* _PyImmutability_Freeze(PyObject* obj)
+                continue;
+            }
+
+            if (PySet_Contains(state->freezable_types, item) == 1) {
+                _Py_SetImmutable(item);
+                continue;
+            }
```
Could a freezable_type have non-trivial state hanging off it, and thus this would prevent that state from being frozen?  Should we still call tp_traverse on this to get that state.

Also, you seem to be calling SetImmutable a lot of times.

matajoh committed:
_datetime

Signed-off-by: Matthew A Johnson <matjoh@microsoft.com>

matajoh committed:
_struct

Signed-off-by: Matthew A Johnson <matjoh@microsoft.com>

matajoh committed:
Making freezable types a weakset

Signed-off-by: Matthew A Johnson <matjoh@microsoft.com>

matajoh committed:
These don't need to be freezable

Signed-off-by: Matthew A Johnson <matjoh@microsoft.com>

mjp41 committed:
Make mutable during finalisation. (#6)

Signed-off-by: Matthew A Johnson <matjoh@microsoft.com>

matajoh committed:
Dealing with more finalisation bugs

Signed-off-by: Matthew A Johnson <matjoh@microsoft.com>

matajoh committed:
Shadow cell

Signed-off-by: Matthew A Johnson <matjoh@microsoft.com>

matajoh committed:
Removing isfreezable and adding some TODOs

Signed-off-by: Matthew A Johnson <matjoh@microsoft.com>

matajoh committed:
Using import helper

Signed-off-by: Matthew A Johnson <matjoh@microsoft.com>

xFrednet commented on Include/internal/pycore_interp.h:
```diff
@@ -176,6 +177,7 @@ struct _is {
     struct _Py_async_gen_state async_gen;
     struct _Py_context_state context;
     struct _Py_exc_state exc_state;
+    struct _Py_immutability_state immutability;
```
Does this mean that different interpreters have different views on what is freezable?

xFrednet commented on Include/object.h:
```diff
@@ -254,16 +273,44 @@ static inline int Py_IS_TYPE(PyObject *ob, PyTypeObject *type) {
         return;
     }
-    ob->ob_refcnt = refcnt;
+    ob->ob_refcnt = (ob->ob_refcnt & _Py_IMMUTABLE_MASK) | (refcnt & _Py_REFCNT_MASK);
```
Shouldn't the immutability flag always be 0 here due to the if statement above?

xFrednet commented:
@matajoh Do we want to change the title of the PR to be ready for outside viewers? Maybe something like "Adding Deep Immutability to 3.12"

xFrednet commented on Modules/_elementtree.c:
```diff
@@ -1730,13 +1757,24 @@ _elementtree_Element_set_impl(ElementObject *self, PyObject *key,
 
+/* macro for writable validation */
+#define _VALIDATE_WRITABLE(op)          \
+    if (!Py_CHECKWRITE(op)) {            \
+        PyErr_SetObject(                 \
+            PyExc_TypeError,      \
+            (PyObject *)((op)));                       \
+        return -1;                     \
+    }
```
Why does this create an Error object and doesn't use `PyErr_WriteToImmutable` like the additions in the other files?

Also why is it called `_VALIDATE_WRITABLE` here and `_CHECK_IS_WRITABLE` above? The difference seems to be the return value, but they seem to be used in the same way.

xFrednet commented on Modules/_elementtree.c:
```diff
@@ -4315,15 +4360,15 @@ static PyMethodDef _functions[] = {
-    if (type == NULL) {                                                  \
-        goto error;                                                      \
-    }                                                                    \
+#define CREATE_TYPE(module, type, spec, base) \
```
Is this new `base` parameter still needed? It seems to be always `NULL` in the uses below

xFrednet commented on Modules/_struct.c:
```diff
@@ -2503,6 +2503,9 @@ _structmodule_exec(PyObject *m)
     if (PyModule_AddType(m, (PyTypeObject *)state->PyStructType) < 0) {
         return -1;
     }
+    if (_PyImmutability_RegisterFreezable((PyTypeObject *)state->PyStructType) < 0){
+        return -1;
+    }
```
The file `Modules/_elementtree.c` uses `_PyImport_GetModuleAttrString("immutable", "register_freezable");` to get the `register_freezable` function or handle failure if the "immutable" module is not available. This one uses the builtin function `_PyImmutability_RegisterFreezable` from pycore. How is it decided which function is used?

xFrednet commented on Modules/cjkcodecs/multibytecodec.c:
```diff
@@ -589,6 +589,12 @@ _multibytecodec_MultibyteCodec_encode_impl(MultibyteCodecObject *self,
     PyObject *errorcb, *r, *ucvt;
     Py_ssize_t datalen;
 
+    if(self->codec->encinit != NULL){
+        if(!Py_CHECKWRITE(self)){
+            return PyErr_WriteToImmutable(self);
+        }
+    }
```
This file has new `Py_CHECKWRITE` but doesn't register the type as freezable. Is this declared somewhere else?

xFrednet commented on Modules/clinic/immutablemodule.c.h:
```diff
@@ -0,0 +1,37 @@
+preserve
+[clinic start generated code]*/
+
+#if defined(Py_BUILD_CORE) && !defined(Py_BUILD_CORE_MODULE)
+#  include "pycore_gc.h"            // PyGC_Head
+#  include "pycore_runtime.h"       // _Py_ID()
+#endif
```
These two includes don't seem to be used in the header file, can we move them into the .c file instead?

xFrednet commented on Modules/immutablemodule.c:
```diff
@@ -0,0 +1,197 @@
+#include "clinic/immutablemodule.c.h"
+
+typedef struct {
+    PyObject *not_freezable_error_obj;
```
Is this object still used? Looking at the code, it seems to only be initialized and cleared, but never used outsite of this

xFrednet commented on Objects/abstract.c:
```diff
@@ -34,6 +34,17 @@ null_error(void)
     return NULL;
 }
 
+static PyObject *
+immutable_error(void)
+{
+    PyThreadState *tstate = _PyThreadState_GET();
+    if (!_PyErr_Occurred(tstate)) {
+        _PyErr_SetString(tstate, PyExc_TypeError,
+                         "cannot modify immutable instance");
+    }
+    return NULL;
+}
```
Why does this used a custom `immutable_error` function and not `PyErr_WriteToImmutable()`. The difference seems to be that this function doesn't take the object that caused the error. But looking at the `Py_CHECKWRITE(o)` we always seem to have an object

xFrednet commented on Python/immutability.c:
```diff
@@ -0,0 +1,588 @@
+        return -1;
+    }
+
+    if(PyDict_SetItemString(PyModule_GetDict(frozen_importlib), "_freezable_types", state->freezable_types)){
+        Py_DECREF(frozen_importlib);
+        return -1;
+    }
```
My understanding is, that this exposes the `freezable_types` set via the `frozen_importlib` module. Is there a reason for this?

xFrednet commented on Python/immutability.c:
```diff
@@ -0,0 +1,588 @@
+    size = PySequence_Fast_GET_SIZE(f_code->co_consts);
+    for(Py_ssize_t i = 0; i < size; i++){
+        PyObject* value = PySequence_Fast_GET_ITEM(f_code->co_consts, i);
+        if(check_globals && PyUnicode_Check(value)){
```
Couldn't we move the `check_globals` check outside the for loop?

xFrednet commented on Python/immutability.c:
```diff
@@ -0,0 +1,588 @@
+                goto error;
+        }
+
+        if(_Py_IsImmutable(item)){
+            continue;
+        }
```
Can't we move this check above `check_freezable`? It seems like it's way cheaper.

matajoh committed:
Adding the bases tuple as discussed

Signed-off-by: Matthew A Johnson <matjoh@microsoft.com>

alwaysmpe commented:
I've been playing with the current implementation. Seems that it’s sometimes trying to freeze type variables that are arguments to generic functions. And that can’t be done. A simple decorator example:
```python
from __future__ import annotations
from immutable import freeze
from typing import Callable
def foo[**P, R](fn: Callable[P, R]) -> Callable[P, R]:
    def fn_wrap(*args: P.args, **kwargs: P.kwargs) -> R:
        return fn(*args, **kwargs)
    return fn_wrap
freeze(foo) # TypeError: Cannot freeze instance of type typing.ParamSpec due to custom functionality implemented in C
```
I’m surprised that `from __future__ import annotations` doesn’t make a difference.

mjp41 commented:
> I've been playing with the current implementation. Seems that it’s sometimes trying to freeze type variables that are arguments to generic functions. And that can’t be done. A simple decorator example:
> 
> ```python
> from __future__ import annotations
> from immutable import freeze
> from typing import Callable
> def foo[**P, R](fn: Callable[P, R]) -> Callable[P, R]:
>     def fn_wrap(*args: P.args, **kwargs: P.kwargs) -> R:
>         return fn(*args, **kwargs)
>     return fn_wrap
> freeze(foo) # TypeError: Cannot freeze instance of type typing.ParamSpec due to custom functionality implemented in C
> ```
> 
> I’m surprised that `from __future__ import annotations` doesn’t make a difference.

Thanks for raising this.  I'll investigate what is happening, and get back to you.

Commit:
Initial simple Python module.

Commit:
Add .clang-format file.

Source: <https://discuss.python.org/t/pep-7-compatible-clang-format-config/66139>
